src/utils/mod.rs

```diff
@@ -705,13 +990,15 @@ impl TryFrom<Signature> for TPMT_SIGNATURE {
             AsymSchemeUnion::ECSchnorr(_) => {
                 Err(Error::local_error(WrapperErrorKind::UnsupportedParam))
             }
-            AsymSchemeUnion::RSAES => Err(Error::local_error(WrapperErrorKind::UnsupportedParam)),
-            AsymSchemeUnion::RSAOAEP(_) => {
-                Err(Error::local_error(WrapperErrorKind::UnsupportedParam))
-            }
             AsymSchemeUnion::AnySig(_) => {
                 Err(Error::local_error(WrapperErrorKind::UnsupportedParam))
             }
+            AsymSchemeUnion::ECDH(_)
+            | AsymSchemeUnion::RSAES
+            | AsymSchemeUnion::RSAOAEP(_)
+            | AsymSchemeUnion::ECMQV(_) => {
+                Err(Error::local_error(WrapperErrorKind::InconsistentParams))
```

Comment:
Should that be `UnsupportedParam`?

Reply:
This is in a `impl TryFrom<Signature> for TPMT_SIGNATURE ` - meaning that only asymmetric schemes that are for signing make sense here; if the scheme on the `Signature` is an encryption one then there's an inconsistency.

That does, however, bring the question of whether the `AsymSchemeUnion` should be split into multiple enums - `AsymSignScheme` and `AsymEncryptScheme`, maybe even further, `RsaSignScheme`, `EccSignScheme` etc.

Reply:
> This is in a impl TryFrom<Signature> for TPMT_SIGNATURE - meaning that only asymmetric schemes that are for signing make sense here; if the scheme on the Signature is an encryption one then there's an inconsistency.

That makes sense!

> That does, however, bring the question of whether the AsymSchemeUnion should be split into multiple enums - AsymSignScheme and AsymEncryptScheme, maybe even further, RsaSignScheme, EccSignScheme etc.

That would mean that the `scheme` field of `Signature` could be `AsymSignScheme` so you would not even have to do that check?

Reply:
> That would mean that the scheme field of Signature could be AsymSignScheme so you would not even have to do that check?

True, but I've also realised that this would only work if we keep separate methods for creating/using signing and decryption keys. Otherwise in the `TransientKeyContext`, for example, we'd have to have different params types or variants for signing and encryption. 

I'm not a huge fan of it, but we can keep the thing as is and simply have these checks here - we have to do a full `match` in most cases anyway. I could implement methods `is_signing` and `is_decryption` on `AsymSchemeUnion` to make checks faster.